core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1100,33 +1214,90 @@ class ReplicaManager(val config: KafkaConfig,
   }
 
   private def getLeaderPartitions: List[Partition] =
-    allPartitions.values.filter(_.leaderReplicaIfLocal.isDefined).toList
+    allPartitions.values.filter(partition => partition != ReplicaManager.OfflinePartition && partition.leaderReplicaIfLocal.isDefined).toList
 
   def getLogEndOffset(topicPartition: TopicPartition): Option[Long] = {
-    getPartition(topicPartition).flatMap{ partition =>
-      partition.leaderReplicaIfLocal.map(_.logEndOffset.messageOffset)
+    getPartition(topicPartition) match {
+      case Some(partition) =>
+        if (partition == ReplicaManager.OfflinePartition)
+          throw new KafkaStorageException(s"Failed to get log end offset for $topicPartition because it in offline log directory")
```

Comment:
I think the behavior of throwing the exception is the same as returning None for `GroupMetadataManager`. I am not sure very what is the right behavior for `TransactionStateManager` if the corresponding partition is in offline log directory. Previously I think it is reasonable to throw KafkaStorageException if the partition is offline. But I am not strong on this. I have changed it to return None instead.